### ISSUE TITLE
CORSMiddleware doesn't work as expected

### ISSUE THREAD

**Issue by harunyasar** [tiangolo/fastapi#5207]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from fastapi import FastAPI
from fastapi.middleware.cors import CORSMiddleware

from fastapi import Request, Response, status


ALLOWED_ORIGIN = r"https?:\/\/.*\.example\.(?:com|cloud)"

app = FastAPI()
app.add_middleware(CORSMiddleware, allow_methods=["GET", "PATCH", "POST"], allow_origin_regex=ALLOWED_ORIGIN)


@app.get("/allowed")
async def cors_get():
    return {"message": "GET allowed for /allowed"}


@app.post("/allowed")
async def cors_post():
    return {"message": "POST allowed for /allowed"}


@app.put("/disallowed")
async def cors_put():
    return {"message": "PUT disallowed for /disallowed"}
```

### Description

I might be confused by the explanation of the [Access-Control-Allow-Methods](https://developer.mozilla.org/en-US/docs/Web/HTTP/CORS#access-control-allow-methods) response header. Because what I expected to see as a result of the preflight request was whether the target resource would be able to respond to any request. According to the example application code above, when the preflight request is made to the `/allowed` endpoint as below, `400 Bad Request` should be returned. Because the target source does not implement a `PATCH` method. But on the contrary, `CORSMiddleware` says this `PATCH` method is allowed. I am also experiencing a similar situation with the `/disallowed` endpoint.

```bash
#### This should have returned "Disallowed CORS method" because PATCH is not implemented
curl -X OPTIONS -H "Origin: http://subdomain.example.com" -H "Access-Control-Request-Method: PATCH"  http://localhost:3000/allowed
OK
```

```bash
#### This should have returned "Disallowed CORS method" because PATCH is not implemented
curl -X OPTIONS -H "Origin: http://subdomain.example.com" -H "Access-Control-Request-Method: PATCH"  http://localhost:3000/disallowed
OK
```

### Operating System

macOS

### Operating System Details

macOS Monterey
Version 12.5 (21G72)

MacBook Pro (16-inch, 2021)
Chip Apple M1 Pro
Memory 32 GB

### FastAPI Version

0.75.0

### Python Version

Python 3.10.0

**Comment by JarroVGIT** [comment]
>  ```
> #### This should have returned "Disallowed CORS method" because PATCH is not implemented
> curl -X OPTIONS -H "Origin: http://subdomain.example.com" -H "Access-Control-Request-Method: PATCH"  http://localhost:3000/allowed
> OK
> ```

`CORSMiddleware` does not look if it is implemented or not, it looks if it is **allowed** or not. And `PATCH` is allowed explicitly in your `CORSMiddleware` configuration. And that is ok, as CORS in itself is about wether or not a request is _allowed_ and not wether or not a request is _valid_.

There is also a technical limitation that prevents `CORSMiddleware` to look a implemented routes. That is because  `CORSMiddleware` is basically an ASGI application in itself, and is just calling the next application in which it is chained. That means it has no notion of what routes are implemented or not (or even that the next ASGI app it is calling has a notion of routes at all). So, this self-made preflight request actually returns correctly.

**Comment by harunyasar** [comment]
@JarroVGIT makes sense. I guess the term resource in the MDN led a confusion on my side. I believe that what is referred to as a resource here is not a specific resource in the REST context, but the server itself.
